### DISCUSSION TITLE
kriged grid with the style 'masked'

### DISCUSSION THREAD

**Discussion by OusBNH** [GeoStat-Framework/PyKrige#209] [Q&A]
Hello and Thanks for your tool,
I am trying to use the style 'masked' for a basin but i don't understand  how i can use the Parameters mask.
```python
import geopandas as gpd
from pykrige.uk import UniversalKriging
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

delimitation_shp = gpd.read_file(r"C:\Users\oboun\zzzzzz\stage\sig_complement\Délimitation_administrative_des_ABH.shp")
bassin1 = delimitation_shp[delimitation_shp["OBJECTID"]==5]
xmin,ymin,xmax,ymax = bassin1.total_bounds
xmin-=0.1
ymin-=0.1
xmax+=0.1
ymax+=0.1
data = np.genfromtxt(r"C:\Users\zzzzzz\Desktop\stage\sig_complement\CSVTOTAL.csv", delimiter=',', skip_header=1)
data = data[data[:,2]>=xmin]
data = data[data[:,2]<=xmax]
data = data[data[:,1]>=ymin]
data = data[data[:,1]<=ymax]

gridx = np.arange(xmin, xmax, 0.06)
gridy = np.arange(ymin,ymax, 0.06)

UK = UniversalKriging(
    data[:, 2],
    data[:, 1],
    data[:, 3],
    variogram_model="spherical",
)

z, ss = UK.execute("masked",gridx, gridy, mask=)
```
so the objectif is to calculate the uk just inside the basin.
![bassin.PNG](https://user-images.githubusercontent.com/30392156/126408588-cd097c0c-2b05-4930-bc75-3d920157378c.PNG)

**Comment by MuellerSeb** [comment]
Hey there,

From the [documentation](https://geostat-framework.readthedocs.io/projects/pykrige/en/stable/generated/pykrige.uk.UniversalKriging.html#pykrige.uk.UniversalKriging.execute):
> mask (boolean array, shape (M, N), optional) – Specifies the points in the rectangular grid defined by xpoints and ypoints that are to be excluded in the kriging calculations. Must be provided if style is specified as ‘masked’. False indicates that the point should not be masked, so the kriging system will be solved at the point. True indicates that the point should be masked, so the kriging system should will not be solved at the point.

This means, you need to write a routine, that generates the mask from your given shape `bassin1`. I don't know `geopandas` enough, but your routine should looks something like this:
```python
def get_mask_from_grid(x, y, shp):
    mask = np.zeros((x.size, y.size), dtype=bool)
    xx, yy = np.meshgrid(x, y, indexing="ij")
    print(mask.shape, xx.shape, yy.shape)
    mask = points_in_shape(shp, xx, yy)  # some special routine form geopandas
    return mask
```
This `points_in_shape` needs to be replace by something special from geopandas. Have a look here: https://stackoverflow.com/questions/48097742/geopandas-point-in-polygon

In summary you need to provide an array with bools that state which point in the grid is masked (outside of your domain).

Does this help?
Cheers, Sebastian
